Support querying and updating UniFi Networks

KyleBrandon opened the pull request:
Added INetwork.ts to provide a contract for the Network entity 
Added Network.ts to represent a network in UniFi
Added Networks.ts to allow for querying all existing networks 
Updated Controller.ts to provide API to get Networks 
Updated Site.ts to provide access to a Networks
Added unit tests for Network.ts
Added unit tests for Networks.ts

## :memo: Description
The following added support for querying the Networks that are available on UniFi gateways.  It provides a Networks object to query the list of networks that are available and a Network object to see the attributes and update an existing network.

The motivation was to provide a mechanism to integrate with Apple HomeKit via Homebridge.  This will allow me to provide a list of networks in the Apple Home app that an be enabled or disabled as needed.  Certain household members do not always stop using the internet when asked, this will allow me the ability to manually disable and to schedule via HomeKit integrations.

### :dart: Relevant issues
It currently exposes all network types the same.  An enhancement would be to create separate types for the Internet, VPN, guest networks, etc.  

### :gem: Type of change

- [x] New feature (non-breaking change which adds functionality)

### :scroll: Example code
```js
  await controller.login();
  const [site] = await controller.sites.list();
  const networks = await site.networks.list();
  const [network] = networks.filter(
    (network) => network.name === "Children Network",
  );
  network.enabled = false;
  await network.save();
``` 

## :vertical_traffic_light: How Has This Been Tested?
### Manual Testing
Testing performed in the form of the example code were used during development.

### Unit Tests
#### Networks::list
**it should list networks**
Mock a list of networks and ensure that calling list() returns the populated Network objects.

**should handle no results**
Mock an empty list of networks and ensure that calling list() returns an empty list.

#### Network::construct::network
**should construct Network with enabled**
Ensure you can create a Network object with a name and 'enabled' equals true.

**should refuse a Network without name**
Ensure you cannot create a Network object without a name.

**should construct with all values**
Ensure you can create a Network object with all values set.

#### Network::functions::save
**it should put the data**
Ensure an existing (mocked) Network object can be updated.

## :checkered_flag: Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] **I have added tests that prove my fix is effective or that my feature works**
- [x] **New and existing unit tests pass locally with my changes**
         _Note that I am unable to get some of the existing tests to run locally_
- [x] I have commented my code, particularly in hard-to-understand areas

KyleBrandon committed:
Support querying and updating UniFi Networks

Added INetwork.ts to provide a contract for the Network entity
Added Network.ts to represent a network in UniFi
Added Networks.ts to allow for querying all existing networks
Updated Controller.ts to provide API to getNetworks
Added unit tests for Network.ts
Added unit tests for Networks.ts

KyleBrandon committed:
Support querying and updating UniFi Networks

Added INetwork.ts to provide a contract for the Network entity
Added Network.ts to represent a network in UniFi
Added Networks.ts to allow for querying all existing networks
Updated Controller.ts to provide API to getNetworks
Added unit tests for Network.ts
Added unit tests for Networks.ts

KyleBrandon commented:
@thib3113 I've updated the PR to include unit tests.  I was able to run these successfully by running: 
`npm run test ./tests/units/Networks.test.ts`
`npm run test ./tests/units/Network.test.ts`.

thib3113 committed:
update gitignore

thib3113 committed:
update gitignore

thib3113 committed:
2FA don't work on non unifiOS (#639)

* upgrade deps

* fix(CI): repair snyk CI

* fix(deps): fix dependencies

* fix(deps): include lock file

* unifi-client-638 2FA don't work on non unifiOS

* follow sonar recommendation

KyleBrandon commented:
@thib3113 Hello, just checking in to see what my next steps are to get this merged.

Kind regards,
Kyle

thib3113 commented:
Hello @KyleBrandon ... In fact, I spend the week end to try to improve the scenario tests ... but I figured that nock can't be a good solution (lot of things that are not correctly handled, or to do manyally) ... 

So, for the moment, I will skip scenarios tests ... I will do some more work on next week end to merge this :) .

KyleBrandon commented:
Great!  Let me know if there's anything I can do to help.  I have another PR queued up for supporting querying the system information and second in the works.

Kyle

thib3113 committed:
Migrate npm (#642)

* save

* disable scenarios tests for the moment

thib3113 committed:
move security snyk check on weekly actions

thib3113 commented on src/Controller.ts:
```diff
@@ -43,10 +44,16 @@ export class Controller extends ObjectWithPrivateValues implements IController {
+
+    get networks(): Networks {
+        this.needLoggedIn();
+        return this._networks;
```
why an object networks on the controller ? 

( this is a "site" object ? not available on the controller level ? )

KyleBrandon replied:
You're right.  I missed that when doing the initial implementation.  I was trying to model my changes after your Site changes originally.  Then I realized that the Networks were more like Devices.  I'll remove the _networks_ from the _Controller_.

thib3113 commented on src/Networks/Network.ts:
```diff
@@ -0,0 +1,312 @@
+        return this._updateDevice(network);
+    }
+
+    protected async _updateDevice(payload: unknown): Promise<this> {
```
please type the payload . 

And why "_updateDevice" ? maybe better to use _updateNetwork ? ( or just update ? )

KyleBrandon replied:
Hold over from my implementation by basing it off your work on _Device_.  I'll fix that up.

thib3113 commented:
I've done some updates on main ( migrating to npm, because problems with yarn ), please update the PR . 

About the rest, it seems my UI doesn't use the same URLs, how did you get the urls ? ( it seems that your URLs seems to works on my setup too . So great ) . 

What are you using to test ? ( unifi consoles ? or not ? )

KyleBrandon commented:
I'm using a UDM Pro for testing on my home network.  I inspected traffic from the web client as I was browsing to and from the Settings > Networks page.  Interestingly I found that it doesn't always query via the _/networkconf_ URL but I just chalked that up to caching.

thib3113 commented:
ok . up to date ? ( I have Network 7.5.165 on mine ) . 

About URL ... unifi looooove to change the url, and how they works ... sometimes it's rest, sometimes not ... I think, they're migrating to a new way to handle rest call ( that's why when testing, I saw an URL starting with /v2 )

KyleBrandon commented:
I'm on UniFi OS 3.1.15 and Network is 7.4.162.
I'm on the official release channel and everything is current.

I've seen v2 URLs cropping up but this seems to be a legacy one.  As mentioned, I found it while inspecting traffic via browser navigation.  I think they heavily cache items as I don't see it every time I navigate to the Networks page.

Kyle

KyleBrandon committed:
Address code review feedback for PR #641

Update Controller and IController to remove exposure of Networks
- Networks are based on a Site and should not be exposed at that level

Update Network
- Renamed `save()` to `setEnabed()` to better describe the intent
- Renamed `_updateDevice()` to `_update()` as this does not update a device and should be generic.  Note that this is a protected method only called by `setEnabled()` currently.

KyleBrandon committed:
Merge branch 'main' into support-unifi-networks

thib3113 commented on src/IController.ts:
```diff
@@ -2,6 +2,7 @@ import { AxiosInstance } from 'axios';
 import { IUser } from './User';
 import { Site, Sites } from './Sites';
 import { UnifiAuth } from './UnifiAuth';
+import { Networks } from './Networks';
```
useless import ?

thib3113 commented on src/Networks/Network.ts:
```diff
@@ -0,0 +1,311 @@
+        this.debug('setEnabled()');
+
+        const network: Partial<INetwork> = {};
+        network.enabled = enabled;
```
```suggestion
        const network: Partial<INetwork> = {
          enabled: true
        };
```

thib3113 commented:
Why did you remove the save() function ? ( I prefer a save function, allowing to update multiples things on the network, than just the enabled = true / false ) . 

You can also create a public function `update` that will allow you to directly pass the enabled = true/false . 

( if your are checking examples, it's better to check FWRule / FWGroup . Devices have lot of differents classes, and that's why _update is protected )

KyleBrandon commented:
When fixing save() I realized that I only tested updating enabled.  I chose to not expose update as I'm not certain what can and can't be updated and with all the settings there's a lot that may not work correctly.
I also had the idea of refactoring Network into something similar with Device to support separate network types of WAN, Corporate, and VPN.  Each of these have different fields that assignable.  I didn't go that way from the start as this is my first Typescript foray contributing to an open source project.  So I appreciate your patience 😄 .

I'm happy to restore the save() if you like, but was leaning towards safety.  Would you prefer to standardize on save() as in FWRule?

Kyle

thib3113 commented:
@KyleBrandon no problems . take your time :) . 

About `save` and `update` . 

I think it can be good that each objects has an update function that allow to update some things (like, I can, skip the loading step, and directly update a network if I know the correct _id) : 
```typescript
const network = new Network();

await network.update({enabled:true});
``` 

and a `save` option that can allow this :
```typescript
const network = new Network();

network.updated = true;

await network.save();
```

( about the test, just check if re-sending the received object works or not ... if yes, it can be fine for the moment ) . 

---

writing this, I remark that only name is mandatory ? but you use _id to do the update ? so, why name and not _id ?

( also, I think one day I'll need to review some mechanism of this library ... but not the time actually ) .

--- 

And about networks separated by types . Yes it can be a solution, to be more precise . 
Depend on the time you want to spend on it :) (if you do it, it can improve request size, readability, and speed)

KyleBrandon committed:
Address code review feedback for PR #641

- Remove unused reference from IController (lint is your friend)
- Update Network to require _id as opposed to name.  The reason being that the _id is required to update or save a network with UniFi, not the name.
- Update Network to support update() by taking a Partial<INetwork> requiring an _id.
- Update Network to clean up the save() method.
- Update Network.test
        - add support method expectNetworkEqual()
        - update construct::network tests to expect _id vs name
        - update functions::save tests to expect _id
        - update functions::update to test updating a network with only an id (no reason to query before)

KyleBrandon commented:
Here are the requested changes.  I've added support for save() as well as update().  I refactored so that _id was required vs. name.  Reason being that it's needed to save or update via the REST API.
Added some tests to cover both save and update as well.

thib3113 commented on src/Networks/Network.ts:
```diff
@@ -0,0 +1,323 @@
+    }
+
+    public async update(props: Partial<INetwork>): Promise<void> {
+        if (!Validate.isDefinedNotNull(props._id)) {
```
```suggestion
        if (!Validate.isDefinedNotNull(this._id)) {
```
not sure what you want to do here ?

KyleBrandon replied:
With update() you'd requested the ability to update with out querying.  So my thought was that update() would need to have the _id passed in as part of the properties to update.  However after reviewing you can't create a Network without an id and we're more interested in ensuring it's available on the Network it self consistently.

Good catch.

thib3113 replied:
@KyleBrandon if you want to do this, it's better to do this with a static method / or a method on "networkS" 

There is no "sense" to get a network, and use it to update another (I think) ?

KyleBrandon replied:
I'm good with the change.  My thinking was flawed.

thib3113 commented on src/Networks/Network.ts:
```diff
@@ -0,0 +1,323 @@
+        return (
+            await this.instance.put('/rest/networkconf/:id', props, {
+                urlParams: {
+                    site: this.site.name,
```
```suggestion
```
site not needed ( added automatically by parent site )

KyleBrandon replied:
Got it

thib3113 commented on src/Networks/Network.ts:
```diff
@@ -0,0 +1,323 @@
+                    id: this._id
+                }
+            })
+        ).data?.data;
```
on my side I change this to : 
```typescript
        await this.instance.put('/rest/networkconf/:id', props, {
            urlParams: {
                id: this._id
            }
        });

        this.import(props);
```

so, the current changes are refleted in the object . 
```typescript
    const networks = await site.networks.list();
    const network = networks.find((n) => n.name === 'test');

    if (!network) {
        throw new Error('not found');
    }

    console.log(network.enabled); // => true

    await network.update({ enabled: false });
    
    console.log(network.enabled); // => false
```

KyleBrandon replied:
Okay, I misunderstood the requirements there.  In your initial request you'd wanted update() to be able to update the Network without requiring querying. 

> like, I can, skip the loading step, and directly update a network if I know the correct _id

This was why I had the method return void.  If you allow a partial to be passed into update then perform `this.import(props);` at the end of the service call Network would be poorly formed with only partial information.

Now I could do what I'm doing in save() and take the return from UniFi which looks to be the entire Network object and import that.  That way update would be fully formed (at least until UniFi changes their API).

Thoughts?

thib3113 replied:
@KyleBrandon the problem here, is in my test, the update return nothing ... 

with this, it will update the current object ( so, when I do the list, the object if filled ... and when I update, the import will only update the updated props )

```typescript
const networks = await site.networks.list();
const network = networks.find((n) => n.name === 'test');


console.log(network);
/**
{
"name": "test",
...
"enabled": true
}
*/

await network.update({ enabled: false });

console.log(network);
/**
{
"name": "test",
...
"enabled": false
}
*/

```

KyleBrandon replied:
Fair point.  The one benefit of importing the properties returned by UniFi is that you would then have a consistent representation from UniFi if something was changed in another method (the UI for instance).

thib3113 replied:
yes yes . I agree ... the problem was, on my side, the update return an empty array ... it's different on your side ?

KyleBrandon replied:
No, that's correct, it was how I was implementing. I was waffling between having update() modify the Network state or not.  
Here's what I propose:

- I'll modify update() to import the result from UniFi, that way we have a consistent state with the gateway.  
- I also propose adding a Validate.isStringEmpty() to verify the id is not undefined, null or empty.  I could do isIdEmpty but I don't know all the validation rules around a UniFi network id.  If you don't like it on Validate I can add it to Network.

Does that sound reasonable?

thib3113 replied:
I'm not sure about the second part ... you want to pass an _id to update another network ? 

about the first one, yes if you can retrieve only the current network ? 

-- 
after some tests, I found : 
```
        const res = await this.instance.get('/rest/networkconf/:id', {
            urlParams: {
                id: this._id
            }
        });
```
so, we can call it to import ? 

something like : 
```
    public async load(): Promise<this> {
        const res = (
            await this.instance.get('/rest/networkconf/:id', {
                urlParams: {
                    id: this._id
                }
            })
        ).data;

        if (res.data) {
            this.import(res.data);
        }

        return this;
    }
```

( and so, you can call load at the end of the update ? )

KyleBrandon replied:
Yes on the update() returning this and importing the response.

Regarding the second part, I was concerned about the _id getting changed or being and empty string.  I'm considering the following:

Add the following to Network:
```
    isIdValid(id: string): boolean {
        if (id.trim().length === 0) {
            return true;
        }

        return true;
    }
```

Update constructor:
```
  constructor(config: IObjectSubSiteConfig, props: Partial<INetwork> & { _id: string }) {
        super(config);

        if (this.isIdValid(props._id)) {
            throw new ClientError('_id is mandatory for a network.', EErrorsCodes.UNKNOWN_ERROR);
        }

        this.import(props);
}
```


Update update():
```
   public async update(props: Partial<INetwork>): Promise<this> {
        if (props._id !== this._id) {
            throw new Error('Cannot update network with a different _id');
        }

        const res = (
            await this.instance.put('/rest/networkconf/:id', props, {
                urlParams: {
                    id: this._id
                }
            })
        ).data;
        if (res.data) {
            this.import(res.data[0]);
        }

        return this;
    }
```

I think those changes would not allow _id to be manipulated after the Network object is constructed.

thib3113 replied:
hum ... I'm not fan of the idea . 

I think, because the object need _id to be construct, other parts need to start with the fact that _id is always defined . 

So, maybe just something like : 
```typescript
    public async update(props: Partial<Omit<INetwork, '_id'>>): Promise<void> {
        await this.instance.put('/rest/networkconf/:id', props, {
```

KyleBrandon replied:
That works slick.

KyleBrandon replied:
Well, I misunderstood what I was seeing in the logs from Axios.  Turns out this conversation is somewhat moot as UniFi does _not_ return the current Network after a put.  I apologize for the misunderstanding, but I did find the debate valuable to understand your thoughts on the direction of the API.

Here are the current save() and update ():

```
    public async update(props: Partial<Omit<INetwork, '_id'>>): Promise<void> {
        await this.instance.put('/rest/networkconf/:id', props, {
            urlParams: {
                id: this._id
            }
        });
        this.import(props);
    }

    public async save(): Promise<this> {
        if (!Validate.isDefinedNotNull(this._id)) {
            throw new Error('Cannot update network without _id');
        }
        const network: INetwork = { ...this } as INetwork;
        await this.instance.put('/rest/networkconf/:id', network, {
            urlParams: {
                id: this._id
            }
        });
        return this;
    }
```

As and additional indication that _id is required and immutable I marked the property _readonly_ and set it in the constructor.

thib3113 replied:
no problem . 

About the readonly on _id . Yes, it can be a good idea :) . 

but remember that this checks is only typescript ... so, when a javascript user use it ... it can be overrided ( maybe with a getter/setter . or with a [javascript private class field](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) ... but it is maybe too much for this time ) . 

( also, remember that readonly need to be defined in constructor, and not in the import ) . 


---

Another thing I saw, what about moving this.debug initialization to constructor ? ( name is more readable ... but can change )
something like : 
````typescript
const debug = createDebugger('Network');

export class Network extends _ObjectSubSite {
    constructor(config: IObjectSubSiteConfig, props: Partial<INetwork>) {
        super(config);

        if (!props._id) {
            throw new ClientError('_id is mandatory for a network.', EErrorsCodes.UNKNOWN_ERROR);
        }

        this._id = props._id;

        this.debug = debug.extend(this._id);

        this.import(props);
    }
```

KyleBrandon replied:
I'll update the debug assignment.  I'll also add some guards on the update() and save() to ensure the _id exists:
```
if (!this._id) {
    throw new Error...
```

If you're comfortable with the proposed changes, I'll look into the getter/setter and private class fields another time 😄 .  Looks like there's a bit to learn there.

Kyle

thib3113 replied:
@KyleBrandon no problem ( I know this is too much for the moment ;) ) . 

I wait your commit :)

KyleBrandon committed:
Address code review feedback for PR #641

- Update Network to set _id as a readonly property in the constructor
- Moved debug initialization out of import()
- Modified update()
	- Omit the _id from the passed properties
	- Ensure that this._id is valid
	- Merge passed properties into this
- Modifed save()
	- Ensure that this._id is valid
	- Remove merge of response as it's empty
- Modified update unit tests
	- Removed check for a valid _id as it's now ignored
	- Added test that passes _id as properties to ensure it's correctly omitted

thib3113 committed:
try CI permissions
